lumen/ai/agents.py

```diff
@@ -491,159 +517,80 @@ class LumenBaseAgent(Agent):
+            feedback=instruction,
+            errors=errors,
+            **kwargs
         )
-
-        with self.param.update(memory=memory):
-            numbered_text = "\n".join(f"{i:2d}: {line}" for i, line in enumerate(lines, 1))
-            system = await self._render_prompt(
-                "retry_output",
-                messages=messages,
-                numbered_text=numbered_text,
-                language=language,
-                feedback=feedback,
-                **context
-            )
         retry_model = self._lookup_prompt_key("retry_output", "response_model")
-        invoke_kwargs = dict(
-            messages=messages,
+        result = await self.llm.invoke(
+            messages,
             system=system,
             response_model=retry_model,
-            model_spec="edit",
+            model_spec="edit"
         )
-        result = await self.llm.invoke(**invoke_kwargs)
+        spec = load_yaml(apply_changes(lines, result.edits))
+        if view is not None:
+            view.validate_spec(spec)
+        if isinstance(spec, str):
+            return spec
+        yaml_spec = dump_yaml(spec)
+        return yaml_spec
 
-        # Apply line changes and return result
-        return self._apply_line_changes_to_output(
-            lines, result.lines_changes, targeted, original_spec, self._retry_target_keys
-        )
 
-    def _render_lumen(
-        self,
-        component: Component,
-        messages: list | None = None,
-        title: str | None = None,
-        **kwargs,
-    ):
-        async def _retry_invoke(event: param.parameterized.Event):
-            with out.param.update(loading=True):
-                out.spec = await self._retry_output_by_line(event.new, messages, memory, out.spec, language=out.language)
-
-        memory = self._memory
-        retry_controls = RetryControls()
-        retry_controls.param.watch(_retry_invoke, "reason")
-        out = self._output_type(
-            component=component,
-            footer=[retry_controls],
-            title=title,
-            **kwargs
-        )
-        out.param.watch(partial(self._update_spec, self._memory), "spec")
-        if "outputs" in self._memory:
-            # We have to create a new list to trigger an event
-            # since inplace updates will not trigger updates
-            # and won't allow diffing between old and new values
-            self._memory["outputs"] = self._memory["outputs"] + [out]
-        if self.interface is not None:
-            message_kwargs = dict(value=out, user=self.user)
-            self.interface.stream(replace=True, max_width=self._max_width, **message_kwargs)
-        return out
+class SQLInputs(ContextModel):
+
+    data: NotRequired[Any]
+
+    source: Source
+
+    sources: Annotated[list[Source], ("accumulate", "source")]
+
+    sql: NotRequired[str]
+
+    sql_metaset: SQLMetaset
+
+    visible_slugs: NotRequired[set[str]]
+
+
+class SQLOutputs(ContextModel):
+
+    data: Any
```

Comment:
Maybe we can specify a custom type for this. If I'm not mistaken it's either a dataframe or a dict.